source/fem-dem/gls_sharp_navier_stokes.cc

```diff
@@ -370,6 +352,12 @@ void
 GLSSharpNavierStokesSolver<dim>::refinement_control(
   const bool initial_refinement)
 {
+  // We add a post-refinement check to update precalculations only if needed
+  bool triangulation_modified_here          = false;
+  bool triangulation_modified_at_least_once = false;
+  this->triangulation->signals.post_refinement.connect(
+    [&triangulation_modified_here]() { triangulation_modified_here = true; });
```

Comment:
Good usage of boost signals
But why is the triangulatio_modified_at_least_once not connected? I'm not sure I follow what is being done here. You need to comment it more please

Reply:
It is to separate the update_precalculations and remove_superfluous_nodes steps. I added more comments to explain it. Is it clearer this way?